cmd/k8s-operator/spdy-hijacker.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/netip"
+	"strings"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"tailscale.com/client/tailscale/apitype"
+	"tailscale.com/ssh/tailssh"
+	"tailscale.com/tailcfg"
+	"tailscale.com/tsnet"
+	"tailscale.com/tstime"
+	"tailscale.com/util/multierr"
+)
+
+// spdyHijacker implements [net/http.Hijacker] interface.
+// It must be configured with an http request for a 'kubectl exec' session that
+// needs to be recorded. It knows how to hijack the connection and configure for
+// the session contents to be sent to a tsrecorder instance.
+type spdyHijacker struct {
+	http.ResponseWriter
+	s        *tsnet.Server
+	req      *http.Request
+	who      *apitype.WhoIsResponse
+	log      *zap.SugaredLogger
+	pod      string           // pod being exec-d
+	ns       string           // namespace of the pod being exec-d
+	addrs    []netip.AddrPort // tsrecorder addresses
+	failOpen bool             // whether to fail open if recording fails
+}
+
+// Hijack hijacks a 'kubectl exec' session and configures for the session
+// contents to be sent to a recorder.
+func (h *spdyHijacker) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	h.log.Infof("recorder addrs: %v, failOpen: %v", h.addrs, h.failOpen)
+	reqConn, brw, err := h.ResponseWriter.(http.Hijacker).Hijack()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error hijacking connection: %w", err)
+	}
+
+	conn, err := h.setUpRecording(reqConn, brw)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error setting up session recording: %w", err)
+	}
+	return conn, brw, nil
+}
+
+// setupRecording attempts to connect to the recorders set via
+// spdyHijacker.addrs. Returns conn from provided opts, wrapped in recording
+// logic. If connecting to the recorder fails or an error is received during the
+// session and spdyHijacker.failOpen is false, connection will be closed.
+func (h *spdyHijacker) setUpRecording(conn net.Conn, brw *bufio.ReadWriter) (net.Conn, error) {
+	const (
+		// https://docs.asciinema.org/manual/asciicast/v2/
+		asciicastv2 = 2
+	)
+	var wc io.WriteCloser
+	h.log.Infof("kubectl exec session will be recorded, recorders: %v, fail open policy: %t", h.addrs, h.failOpen)
+	ctx := context.Background()
+	rw, _, errChan, err := tailssh.ConnectToRecorder(ctx, h.addrs, h.s.Dialer())
+	if err != nil {
+		msg := fmt.Sprintf("error connecting to session recorders: %v", err)
+		if !h.failOpen {
+			msg = msg + "; failure mode is 'fail closed'; closing connection."
+			if err := closeConnWithWarning(conn, msg); err != nil {
+				return nil, multierr.New(errors.New(msg), err)
+			}
+			return nil, errors.New(msg)
+		} else {
+			msg = msg + "; failure mode is 'fail open'; continuing session without recording."
+			h.log.Warnf(msg)
+			return conn, nil
+		}
+	} else {
```

Comment:
nit: I don't think this `else` is necessary, above we always return early if `err != nil`